File: src/PixUI/PixUI/Platform/Common/Map.cs
using System.Collections;
using System.Collections.Generic;
using System.Diagnostics.CodeAnalysis;
using PixUI;

namespace System
{
    //用于简单转换为Web的Map<string,T> or Map<number,T>

    [TSType("Map")]
    [SuppressMessage("ReSharper", "InconsistentNaming")]
    public abstract class MapBase<K, V>
    {
        private readonly Dictionary<K, V> _dictionary = new Dictionary<K, V>();

        protected MapBase(ValueTuple<K, V>[] items)
        {
            foreach (var entry in items)
            {
                _dictionary.Add(entry.Item1, entry.Item2);
            }
        }

        public V? get(K key)
        {
            _dictionary.TryGetValue(key, out var value);
            return value;
        }

        public void set(K key, V value)
        {
            _dictionary[key] = value;
        }

        public bool has(K key) => _dictionary.ContainsKey(key);

        public bool delete(K key) => _dictionary.Remove(key);

        public ICollection<K> keys => _dictionary.Keys;

        public ICollection<V> values => _dictionary.Values;
    }

    [TSType("System.NumberMap")]
    public sealed class NumberMap<T> : MapBase<int, T>
    {
        public NumberMap(ValueTuple<int, T>[] items) : base(items) { }
    }

    [TSType("System.StringMap")]
    public sealed class StringMap<T> : MapBase<string, T>
    {
        public StringMap(ValueTuple<string, T>[] items) : base(items) { }
    }
}

File: src/Dev.Web/src/System/Collections/Map.ts
import {IEnumerable} from "@/System";
import {from} from "@/System/Linq";

class MapBase<TKey, TValue> extends Map<TKey, TValue> {
    public get Keys(): IEnumerable<TKey> {
        return from(this.keys());
    }

    public get Values(): IEnumerable<TValue> {
        return from(this.values());
    }

    public Add(key: TKey, value: TValue) {
        this.set(key, value);
    }

    public Remove(key: TKey) {
        this.delete(key);
    }

    public Clear() {
        this.clear();
    }
}

export class NumberMap<T> extends MapBase<number, T> {}

export class StringMap<T> extends MapBase<string, T> {}

export class DoubleMap<T> extends MapBase<number, T> {}

export class ObjectMap<T> extends MapBase<any, T> {}